core-rust/node-common/src/locks.rs

```diff
@@ -0,0 +1,212 @@
+    pub fn lock(&self) -> MutexGuard<'_, T> {
+        MutexGuard {
+            underlying: self.underlying.lock(),
+        }
+    }
+}
+
+/// A panic-safe facade for a [`parking_lot::MutexGuard`].
+pub struct MutexGuard<'a, T> {
+    underlying: parking_lot::MutexGuard<'a, T>,
+}
+
+impl<'a, T: 'a> Deref for MutexGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        self.underlying.deref()
+    }
+}
+
+impl<'a, T: 'a> DerefMut for MutexGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.underlying.deref_mut()
+    }
+}
+
+impl<'a, T> Drop for MutexGuard<'a, T> {
+    fn drop(&mut self) {
+        abort_if_panicking(self);
+    }
+}
+
+/// A panic-safe facade for a [`parking_lot::RwLock`].
+pub struct RwLock<T> {
+    underlying: parking_lot::RwLock<T>,
+}
+
+impl<T> RwLock<T> {
+    /// Wraps the given value in a panic-safe [`RwLock`].
+    pub fn new(value: T) -> Self {
+        Self {
+            underlying: parking_lot::const_rwlock(value),
+        }
+    }
+
+    /// Delegates to the [`parking_lot::RwLockReadGuard::read()`], but returns a panic-safe guard.
+    pub fn read(&self) -> RwLockReadGuard<'_, T> {
+        RwLockReadGuard {
+            underlying: self.underlying.read(),
+        }
+    }
+
+    /// Delegates to the [`parking_lot::RwLockReadGuard::write()`], but returns a panic-safe guard.
+    pub fn write(&self) -> RwLockWriteGuard<'_, T> {
+        RwLockWriteGuard {
+            underlying: self.underlying.write(),
+        }
+    }
+}
+
+/// A panic-safe facade for a [`parking_lot::RwLockReadGuard`].
+pub struct RwLockReadGuard<'a, T> {
+    underlying: parking_lot::RwLockReadGuard<'a, T>,
+}
+
+impl<'a, T: 'a> Deref for RwLockReadGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        self.underlying.deref()
+    }
+}
+
+impl<'a, T> Drop for RwLockReadGuard<'a, T> {
+    fn drop(&mut self) {
+        // The impl here is deliberately no-op:
+        // The lock guard will be dropped (i.e. lock released) on its own, and in case of reading,
+        // we do not need to abort the process (since unmodified state means consistent state).
+    }
+}
+
+/// A panic-safe facade for a [`parking_lot::RwLockWriteGuard`].
+pub struct RwLockWriteGuard<'a, T> {
+    underlying: parking_lot::RwLockWriteGuard<'a, T>,
+}
+
+impl<'a, T: 'a> Deref for RwLockWriteGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        self.underlying.deref()
+    }
+}
+
+impl<'a, T: 'a> DerefMut for RwLockWriteGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.underlying.deref_mut()
+    }
+}
+
+impl<'a, T> Drop for RwLockWriteGuard<'a, T> {
+    fn drop(&mut self) {
+        abort_if_panicking(self);
+    }
+}
+
+fn abort_if_panicking<T>(_guard: &T) {
+    if thread::panicking() {
+        error!(
+            "a {} was released while panicking; aborting the process",
+            type_name::<T>()
+        );
+        abort();
```

Comment:
Can't the abort happen while writing a fresh commit to db?

I think it's a bit extreme, we should at least wait for any ongoing write.

Reply:
As discussed on Slack:

Looks like the best alternative to the (indeed extreme) `abort()` is to call a JNI `System.exit()`. (it will relatively-gracefully clean up java and rust resources via our shutdown-hook).
Let me try to achieve it.

Reply:
Alright, I am able to call a `System.exit(-1)` from a `rwGuard.drop()` (instead of `abort()`).

However, this is what happens:
```
thread 'tokio-runtime-worker' panicked at 'some bug', core-api-server/src/core_api/handlers/mempool_list.rs:11:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/39f2657d1101b50f9b71ae460b762d330cc8426b/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/39f2657d1101b50f9b71ae460b762d330cc8426b/library/core/src/panicking.rs:64:14
   2: core_api_server::core_api::handlers::mempool_list::handle_mempool_list::{{closure}}::{{closure}}
             at /Users/jakrawcz/repo1/babylon-node/core-rust/core-api-server/src/core_api/handlers/mempool_list.rs:11:13
   ... // snipped
 127: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /Users/jakrawcz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.28.2/src/runtime/blocking/pool.rs:471:13
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
2023-07-26T11:30:35.900642Z ERROR node_common::locks: a node_common::locks::RwLockWriteGuard<state_manager::mempool::priority_mempool::PriorityMempool> was released while panicking; invoking System.exit()
Node 833d72:03fff9 is shutting down... // our shutdown hook, nice
Pool BFT no...64vp9cx0t did not terminate // whoops
```

... and nothing more, the process hangs:
![image](https://github.com/radixdlt/babylon-node/assets/119413677/ae5ab4f5-819d-4b45-887f-8a830d337dd1)

Indeed, the `BFT no...64vp9cx0t` did not terminate:
<img width="741" alt="image" src="https://github.com/radixdlt/babylon-node/assets/119413677/2440aaeb-cdff-4830-8b63-2c466b882081">

(the java `.shutdownNow()` could not interrupt it, since it hangs inside the Rust commit)
Only the `kill -9` works.

Is there a way to investigate where exactly in `commit()` we hang?
(so far I've learnt that `rust-gdb` is not available on m1)

I suspect waiting for a lock that is held by some thread which did not release it while unwinding... (although it is supposed to). I know it is not the lock on which I triggered the `System.exit(-1)` (my bug uses a `RwLock::new()`).

Reply:
ahhh it's more complicated: depending on timing, sometimes the BFT pool shutdowns fine.

However, in such case, Rust will hang here:

<img width="833" alt="image" src="https://github.com/radixdlt/babylon-node/assets/119413677/07df3221-729d-4a7a-a96c-738c12c3d75e">

(that's our shutdown hook)

That cleanup method only does:
```
let jni_state_manager: JNIStateManager = env
            .take_rust_field(j_state_manager, POINTER_JNI_FIELD_NAME)
            .unwrap();
```
... so it's hanging inside the `jni` lib :(

Reply:
ok, my mad debugging skillz ("add lots of logging") led me to believe that the thread hangs when dropping the `JNIStateManager` instance (it prints a log from inside its `drop()`, but does not print the log from the very next line, consistently).

Still, I have no idea how to progress.

Reply:
For curious: here is my current state, for easy repro of the "app hangs during cleanup after Rust-induced System.exit()". I welcome any help.

The branch:
https://github.com/radixdlt/babylon-node/compare/refactor/abort_on_broken_synchro...experiment/access_jvm_statically?expand=1

1. Run Single Validator.
2. `curl localhost:3333/core/mempool/list -H 'Content-Type: application/json' -d '{"network":"bomb"}' -v`
3. Observe the process shutting down very orderly, but then hanging just before `println!("field dropped");`
4. `jps` + `kill -9 ...` to kill it

Reply:
The "shutdown hook's thread hangs" problem mentioned above can be fixed by triggering an async shutdown; SOLVED 👍 

However, there is another problem: after extensive experimenting with @malexandru-rdx I must conclude that there is no reliable way of statically obtaining a reference to a running JVM across different JVM versions, JNI versions and rustc versions - it can always happen that someone has a set-up on which the JNI native `JNI_GetCreatedJavaVMs()` returns invalid pointer, which results in SIGBUS or SIGSEGV :(

This means that we no longer can have static lock constructors, but rather a managed `LockFactory`, which is configured with an "app-killer" callback (which will do the most regular `JNIEnv.call(async_exit)`).
Maybe it's even more elegant, since e.g. in tests we do NOT want to do `System.exit()` (i.e. we would prefer to have a "test-fail'er" callback there, not "app-killer" - in order to capture proper test results).

Anyways, that's ☝️ my design and I know how to progress (it's just a bit more code changes).

Reply:
> I must conclude that there is no reliable way of statically obtaining a reference to a running JVM

I have managed to fix it, on the `experiment/` branch. I think the problem was a mismatch between multiple JVM libraries.

> This means that we no longer can have static lock constructors, but rather a managed LockFactory, which is configured with an "app-killer" callback (which will do the most regular JNIEnv.call(async_exit)).
Maybe it's even more elegant, since e.g. in tests we do NOT want to do System.exit() (i.e. we would prefer to have a "test-fail'er" callback there, not "app-killer" - in order to capture proper test results).

As discussed on Slack, this is a bit too hacky (or too much unnecessary complication) for Rust side - I'd prefer to not let testing from Java interfere with production/arhitecture in Rust at the expense of Java code's quality.

Reply:
Alright, the newest commit introduced the change requested above - invoking java `System.exit()` instead of `abort`.
(credits to @malexandru-rdx for helping with rust build to include jni-sys properly).

Worth noting: after this change, the "shutdown" is reaaaaaallly asynchronous. (and it's not even because of our async `System.exit()` call - the shutdown hooks themselves are always async by themselves)
This means that theoretically, a couple of BFT rounds can go through before the relevant Java/Rust services are stopped. And a couple of Core API calls too.